src/Microsoft.AspNet.Razor/TagHelpers/TagHelperDescriptorProvider.cs

```diff
@@ -58,23 +59,41 @@ public IEnumerable<TagHelperDescriptor> GetTagHelpers(string tagName)
                 descriptors = new HashSet<TagHelperDescriptor>(TagHelperDescriptorComparer.Default);
             }
 
-            // If the requested tag name is the catch-all target, we should short circuit.
-            if (tagName.Equals(CatchAllDescriptorTarget, StringComparison.OrdinalIgnoreCase))
+            // If the requested tag name is the catch-all target, we shouldn't do the work of concatenating extra
+            // descriptors.
+            if (!tagName.Equals(CatchAllDescriptorTarget, StringComparison.OrdinalIgnoreCase))
             {
-                return descriptors;
+                // If we have a tag name associated with the requested name, we need to combine matchingDescriptors 
+                // with all the catch-all descriptors.
+                HashSet<TagHelperDescriptor> matchingDescriptors;
+                if (_registrations.TryGetValue(tagName, out matchingDescriptors))
+                {
+                    descriptors = new HashSet<TagHelperDescriptor>(matchingDescriptors.Concat(descriptors));
```

Comment:
just `descriptors = matchingDescriptors.Concat(descriptors);`.  should have already eliminated duplicate descriptors.